text/0000-contingency-plan.md

```diff
@@ -0,0 +1,425 @@
+- Stack overflow is currently detected via a segmented stack check
+  prologue and results in an abort. We expect to experiment with a
+  system based on guard pages in the future.
+- We currently abort the process on OOM conditions (exceeding the heap space, overflowing
+  the stack). We may attempt to panic in such cases instead if possible.
+- Some details of type inference may change. For example, we expect to
+  implement the fallback mechanism described in [RFC 213], and we may
+  wish to make minor changes to accommodate overloaded integer
+  literals. In some cases, type inferences changes may be better
+  handled via explicit opt-in.
+
+(Although it is not directly covered by this RFC, it's worth noting in
+passing that some of the CLI flags to the compiler may change in the
+future as well. The `-Z` flags are of course explicitly unstable, but
+some of the `-C`, rustdoc, and linker-specific flags are expected to
+evolve over time.)
+
+### Opt-in changes
+
+For breaking changes that are not related to soundness or language
+semantics, but are still deemed desirable, an opt-in strategy can be
+used instead. This section describes an attribute for opting in to
+newer language updates, and gives guidelines on what kinds of changes
+should or should not be introduced in this fashion.
+
+We use the term *"opt-in changes"* to refer to changes that would be
+breaking changes, but are not because of the opt-in mechanism.
+
+#### Rust version attribute
+
+The specific proposal is an attribute `#![rust_version="X.Y"]` that
+can be attached to the crate; the version `X.Y` in this attribute is
+called the crate's "declared version". Every build of the Rust
+compiler will also have a version number built into it reflecting the
+current release.
+
+When a `#[rust_version="X.Y"]` attribute is encountered, the compiler
+will endeavor to produce the semantics of Rust "as it was" during
+version `X.Y`. RFCs that propose opt-in changes should discuss how the
+older behavior can be supported in the compiler, but this is expected
+to be straightforward: if supporting older behavior is hard to do, it
+may indicate that the opt-in change is too complex and should not be
+accepted.
+
+If the crate declares a version `X.Y` that is *newer* than the
+compiler itself, the compiler should simply issue a warning and
+proceed as if the crate had declared the compiler's version (i.e., the
+newer version the compiler knows about).
+
+Note that if the changes introducing by the Rust version `X.Y` affect
+parsing, implementing these semantics may require some limited amount
+of feedback between the parser and the tokenizer, or else a limited
+"pre-parse" to scan the set of crate attributes and extract the
+version. For example, if version `X.Y` adds new keywords, the
+tokenizer will likely need to be configured appropriately with the
+proper set of keywords. For this reason, it may make sense to require
+that the `#![rust_version]` attribute appear *first* on the crate.
```

Comment:
@sfackler @seanmonstar ah, interesting thought. @alexcrichton @brson @wycats -- thoughts? seems to touch on general cargo/rustc interaction.